pkg/controller/common/helpers.go

```diff
@@ -337,39 +420,46 @@ func ValidateMachineConfig(cfg mcfgv1.MachineConfigSpec) error {
-			if errV3_0 == nil && !rptV3_0.IsFatal() {
-				return translate3.Translate(translate3_1.Translate(ignCfgV3_0)), nil
+		if errV3_2.Error() == ign3error.ErrUnknownVersion.Error() {
+			ignCfgV3_1, rptV3_1, errV3_1 := ign3_1.Parse(rawIgn)
+			if errV3_1 == nil && !rptV3_1.IsFatal() {
+				return translate3.Translate(translate3_2.Translate(ignCfgV3_1)), nil
 			}
-
-			if errV3_0.Error() == ign3error.ErrUnknownVersion.Error() {
-				ignCfgV2, rptV2, errV2 := ign2.Parse(rawIgn)
-				if errV2 == nil && !rptV2.IsFatal() {
-					return ignCfgV2, nil
+			// unlike spec v2 parsers, v3 parsers aren't chained by default so we need to try parsing as spec v3.0 as well
+			if errV3_1.Error() == ign3error.ErrUnknownVersion.Error() {
+				ignCfgV3_0, rptV3_0, errV3_0 := ign3_0.Parse(rawIgn)
+				if errV3_0 == nil && !rptV3_0.IsFatal() {
+					return translate3.Translate(translate3_2.Translate(translate3_1.Translate(ignCfgV3_0))), nil
 				}
 
-				// If the error is still UnknownVersion it's not a 3.2/3.1/3.0 or 2.x config, thus unsupported
-				if errV2.Error() == ign2error.ErrUnknownVersion.Error() {
-					return ign3types.Config{}, errors.Errorf("parsing Ignition config failed: unknown version. Supported spec versions: 2.2, 3.0, 3.1, 3.2")
+				if errV3_0.Error() == ign3error.ErrUnknownVersion.Error() {
+					ignCfgV2, rptV2, errV2 := ign2.Parse(rawIgn)
+					if errV2 == nil && !rptV2.IsFatal() {
+						return ignCfgV2, nil
+					}
+
+					// If the error is still UnknownVersion it's not a 3.2/3.1/3.0 or 2.x config, thus unsupported
+					if errV2.Error() == ign2error.ErrUnknownVersion.Error() {
+						return ign3types.Config{}, errors.Errorf("parsing Ignition config failed: unknown version. Supported spec versions: 2.2, 3.0, 3.1, 3.2")
+					}
+					return ign3types.Config{}, errors.Errorf("parsing Ignition spec v2 failed with error: %v\nReport: %v", errV2, rptV2)
 				}
-				return ign3types.Config{}, errors.Errorf("parsing Ignition spec v2 failed with error: %v\nReport: %v", errV2, rptV2)
+				return ign3types.Config{}, errors.Errorf("parsing Ignition config spec v3.0 failed with error: %v\nReport: %v", errV3_0, rptV3_0)
 			}
-			return ign3types.Config{}, errors.Errorf("parsing Ignition config spec v3.0 failed with error: %v\nReport: %v", errV3_0, rptV3_0)
+			return ign3types.Config{}, errors.Errorf("parsing Ignition config spec v3.1 failed with error: %v\nReport: %v", errV3_1, rptV3_1)
 		}
-		return ign3types.Config{}, errors.Errorf("parsing Ignition config spec v3.1 failed with error: %v\nReport: %v", errV3_1, rptV3_1)
+		return ign3types.Config{}, errors.Errorf("parsing Ignition config spec v3.2 failed with error: %v\nReport: %v", errV3_2, rptV3_2)
 	}
-	return ign3types.Config{}, errors.Errorf("parsing Ignition config spec v3.2 failed with error: %v\nReport: %v", errV3_2, rptV3_2)
+	return ign3types.Config{}, errors.Errorf("parsing Ignition config spec v3.3 failed with error: %v\nReport: %v", errV3_3, rptV3_3)
```

Comment:
This, and to a smaller extent `ConvertRawExtIgnitionToV3`, is getting unreasonable.

Is there any chance to make this a table-driven loop instead?
```go
for _, attempt := range []struct{attempt func([]byte) (interface{}, error)} {
    func([]byte) (interface{}, error) {
        …
    },
    …
} {
   res, err := attempt(rawIgn)
   if err != nil { return err}
   if res != nil { return res, nil}
}
```
or maybe with the attempt functions out of line?

I don’t expect that we can get rid of all the duplication without generics, but at least expressing the top-level structure of this function and getting the error wrappers closer to the error locations should be possible.

Or maybe this just needs inverting the `if errV3_3.Error() == ign3error.ErrUnknownVersion.Error()` checks, so that the function aborts early, and all of this turns into a long repetitive straight-line code, with no nesting and no `func` intermediates.

Reply:
Can we use the new API from https://github.com/coreos/ignition/pull/1193 here?